### ISSUE TITLE
`assertion failed: expected.is_empty()` when compiling

### ISSUE THREAD

**Issue by nrc** [nikomatsakis/lalrpop#276]
This is my first go at using LALRPOP, so I'm probably doing something silly. When I building my project I get the above failed assertion. Backtrace is:

```
6: lalrpop::build::parse_and_normalize_grammar
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/build/mod.rs:212
   7: lalrpop::build::gen_resolve_file
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/build/mod.rs:100
   8: lalrpop::build::process_file
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/build/mod.rs:43
   9: lalrpop::build::process_dir
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/build/mod.rs:34
  10: lalrpop::api::Configuration::process_dir
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/api/mod.rs:146
  11: itertools::adaptors::dedup
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/api/mod.rs:140
  12: lalrpop::api::Configuration::process_dir
             at /Users/nick/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.13.1/src/api/mod.rs:163
  13: build_script_build::main
             at ./build.rs:4
```

There is a single lalrpop file at src/parse_query/query.lalrpop. The contents of that file doesn't seem to change the assertion, unless I delete everything and then I get an 'unexpected EOL`.

**Comment by brendanzab** [comment]
Sometimes I get this when I forget to put in a `;` between rules. Been meaning to report it, but I forget because it's easy to fix for me now! Could you post a gist of the failing grammar?

**Comment by nrc** [comment]
I was missing some semi-colons, but adding them didn't fix this. A mimimal version of my grammar is:

```
use std::str::FromStr;

grammar;

pub Integer: i64 {
    r"[0-9]+" => from_str(<>).unwrap(),
};
```

**Comment by brendanzab** [comment]
Should be this I think:

```rust
use std::str::FromStr;

grammar;

pub Integer: i64 = {
    r"[0-9]+" => from_str(<>).unwrap(),
};
```

**Comment by nrc** [comment]
Yeah, that fixed my minimal case and then I got the same thing again for other issues. So I think this assertion is just the generic 'something is wrong' error and the underlying issue is that error reporting is not fully implemented. Since that is a known issue, may as well close this.

**Comment by brendanzab** [comment]
I dunno, I think it might be worth keeping this open. Or make an issue that notes down places that give obscure errors.

**Comment by nrc** [comment]
OK, re-opened. FTR there were quite a few trivial syntactic issues that caused this assertion (missing semi-colons, missing `=`, etc.) plus what seemed to be more semantic parsing issues (I forget my theory, but I fixed it by refactoring my rules so that there was always distinguishing terminal).

**Comment by masonk** [comment]
I'm here because I had the same error tonight.

I saw this same error in each of the following cases :

1. A use statement was referring to a library that didn't exist.
2. I had a misplaced semicolon.
3. I forgot to annotate the type of a nonterminal that required annotation.
4. The resultant production rules didn't typecheck.

Ultimately the only way I was able to proceed was to start from scratch and add each line back to the file one at a time and recompile.

I got there in the end and it's working well now, (thanks!) but it would be nice to have some more help in the process.

**Comment by brendanzab** [comment]
I brought [this up on gitter](https://gitter.im/lalrpop/Lobby?at=5a60826f290a1f4561a341a4) and according to @nikomatsakis there is currently a fix on master. I think it's just waiting to be deployed to crates.io.

**Comment by nikomatsakis** [comment]
Yeah, I think this is a bug, I plan to issue a new rew release shortly that fixes it and also some other things.

**Comment by nikomatsakis** [comment]
Using @nrc's example:

```
use std::str::FromStr;

grammar;

pub Integer: i64 {
    r"[0-9]+" => from_str(<>).unwrap(),
};
```

I get (on master):

```
../test.lalrpop:5:18: 5:18 error: unexpected token: `{`
  pub Integer: i64 {
                   ^
```

If I add a `=` and remove the final `;`, I get:

```
../test.lalrpop:8:1: 8:0 error: unexpected end of file

  ^
```

```
